api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,758 @@
+			AvailableCPU:     availableCPU,
+			AvailableMemory:  availableMemory,
+			ID:               aws.StringValue(instance.Ec2InstanceId),
+			InUse:            inUse,
+			Status:           aws.StringValue(instance.Status),
+			UsedPorts:        usedPorts,
+		}
+
+		result = append(result, r)
+	}
+
+	return result, nil
+}
+
+func (e *EnvironmentScaler) scaleDown(clusterName string, desiredScale int, asg *autoscaling.Group, unusedProviders []*ResourceProvider) error {
+	minCapacity := int(aws.Int64Value(asg.MinSize))
+	if desiredScale < minCapacity {
+		log.Printf("[DEBUG] [EnvironmentScaler] Minimum capacity is '%d'. Aborting scaling action for environment '%s'.", minCapacity, clusterName)
+		return nil
+	}
+
+	input := &autoscaling.SetDesiredCapacityInput{}
+	input.SetAutoScalingGroupName(aws.StringValue(asg.AutoScalingGroupName))
+	input.SetDesiredCapacity(int64(desiredScale))
+
+	if _, err := e.Client.AutoScaling.SetDesiredCapacity(input); err != nil {
+		return err
+	}
+
+	// choose which instances to terminate during our scale-down process
+	// instead of having asg randomly select instances
+	// e.g. if we scale from 5 to 3, we can terminate up to 2 unused instances
+	currentCapacity := int(aws.Int64Value(asg.DesiredCapacity))
+	maxNumberOfInstancesToTerminate := currentCapacity - desiredScale
+
+	canTerminate := func(i int) bool {
+		if i+1 > maxNumberOfInstancesToTerminate {
+			return false
+		}
+
+		if i > len(unusedProviders)-1 {
+			return false
+		}
+
+		return true
+	}
+
+	for i := 0; canTerminate(i); i++ {
+		unusedProvider := unusedProviders[i]
+		log.Printf("[DEBUG] [EnvironmentScaler] Terminating unused instance '%s' from environment '%s'.", unusedProvider.ID, clusterName)
+
+		input := &autoscaling.TerminateInstanceInAutoScalingGroupInput{}
+		input.SetInstanceId(unusedProvider.ID)
+		input.SetShouldDecrementDesiredCapacity(false)
+
+		if _, err := e.Client.AutoScaling.TerminateInstanceInAutoScalingGroup(input); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (e *EnvironmentScaler) scaleTo(clusterName string, desiredScale int, unusedProviders []*ResourceProvider) error {
+	input := &autoscaling.DescribeAutoScalingGroupsInput{}
+	input.SetAutoScalingGroupNames([]*string{&clusterName})
+
+	asgs, err := e.Client.AutoScaling.DescribeAutoScalingGroups(input)
+	if err != nil {
+		return err
+	}
+
+	asg := asgs.AutoScalingGroups[0]
+
+	currentCapacity := int(aws.Int64Value(asg.DesiredCapacity))
+
+	switch {
+	case desiredScale > currentCapacity:
+		log.Printf("[DEBUG] [EnvironmentScaler] Attempting to scale environment '%s' from current scale of '%d' to desired scale of '%d'.", clusterName, currentCapacity, desiredScale)
+		return e.scaleUp(clusterName, desiredScale, asg)
+
+	case desiredScale < currentCapacity:
+		log.Printf("[DEBUG] [EnvironmentScaler] Attempting to scale environment '%s' from current scale of '%d' to desired scale of '%d'.", clusterName, currentCapacity, desiredScale)
```

Comment:
This is the same amount of context we had as before for a scaling action. I think it would be helpful to have more details --- would it be possible to get the specific instance id about to be scaled down, for instance?

Reply:
@diemonster That information does come through in a later `[DEBUG]` log; check near the end of `scaleDown()`.

Reply:
✅